<tt>
&lt;!DOCTYPE&nbsp;HTML&nbsp;PUBLIC&nbsp;&quot;-//W3C//DTD&nbsp;HTML&nbsp;3.2//EN&quot;&gt;<br>
&lt;HTML&gt;<br>
&lt;HEAD&gt;<br>
&lt;META&nbsp;HTTP-EQUIV=&quot;Content-Type&quot;&nbsp;CONTENT=&quot;text/html;&nbsp;charset=us-ascii&quot;&gt;<br>
&lt;META&nbsp;NAME=&quot;Generator&quot;&nbsp;CONTENT=&quot;MS&nbsp;Exchange&nbsp;Server&nbsp;version&nbsp;6.5.6944.0&quot;&gt;<br>
&lt;TITLE&gt;Re:&nbsp;[c++-pthreads]&nbsp;&nbsp;Re:&nbsp;pthread_cancel&nbsp;and&nbsp;EH:&nbsp;let's&nbsp;try&nbsp;this&nbsp;again&lt;/TITLE&gt;<br>
&lt;/HEAD&gt;<br>
&lt;BODY&gt;<br>
&lt;!--&nbsp;Converted&nbsp;from&nbsp;text/rtf&nbsp;format&nbsp;--&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;Below&nbsp;is&nbsp;my&nbsp;old&nbsp;note&nbsp;to&nbsp;Jakub&nbsp;with&nbsp;description&nbsp;of&nbsp;failure&nbsp;in&nbsp;my&nbsp;Linux&nbsp;pthread&nbsp;test&nbsp;program&nbsp;which&nbsp;is&nbsp;based&nbsp;on&nbsp;thread&nbsp;pool&nbsp;implementation,&nbsp;i.e.&nbsp;it&nbsp;has&nbsp;to&nbsp;catch&nbsp;exceptions&nbsp;thrown&nbsp;by&nbsp;cancelled&nbsp;worker&nbsp;thread.&nbsp;That&nbsp;illustrates&nbsp;why&nbsp;we&nbsp;could&nbsp;not&nbsp;use&nbsp;Linux&nbsp;pthread.&nbsp;From&nbsp;my&nbsp;point&nbsp;of&nbsp;view&nbsp;this&nbsp;is&nbsp;caused&nbsp;by&nbsp;use&nbsp;of&nbsp;EH&nbsp;to&nbsp;implement&nbsp;thread&nbsp;cancellation.&lt;/FONT&gt;&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;-&nbsp;George&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;-----Original&nbsp;Message-----&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;From:&nbsp;George&nbsp;Shimanovich&nbsp;&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;Sent:&nbsp;Thursday,&nbsp;June&nbsp;24,&nbsp;2004&nbsp;3:52&nbsp;PM&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;To:&nbsp;'Jakub&nbsp;Jelinek'&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;Subject:&nbsp;RE:&nbsp;Bug&nbsp;111548:&nbsp;calling&nbsp;pthread_cancel&nbsp;in&nbsp;a&nbsp;muti-thread&nbsp;c++&nbsp;application&nbsp;abort()s&nbsp;the&nbsp;app&nbsp;-&nbsp;'cancel&nbsp;state'&nbsp;workaround&lt;/FONT&gt;&lt;/P&gt;<br>
&lt;BR&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;Jakub,&lt;/FONT&gt;<br>
&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;I&nbsp;just&nbsp;implemented&nbsp;workaround&nbsp;that&nbsp;you&nbsp;suggested&nbsp;-&nbsp;for&nbsp;deferred&nbsp;threads,&nbsp;-&nbsp;with&nbsp;the&nbsp;same&nbsp;result:&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;1)&nbsp;disable&nbsp;cancel&nbsp;state&nbsp;at&nbsp;the&nbsp;beginning&nbsp;of&nbsp;new&nbsp;thread&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;2)&nbsp;enable&nbsp;it&nbsp;before&nbsp;each&nbsp;call&nbsp;to&nbsp;pthread_testcancel()&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;3)&nbsp;disable&nbsp;cancel&nbsp;state&nbsp;after&nbsp;each&nbsp;call&nbsp;to&nbsp;pthread_testcancel()&nbsp;Just&nbsp;to&nbsp;make&nbsp;sure&nbsp;I&nbsp;explicitly&nbsp;set&nbsp;pending&nbsp;cancel&nbsp;type.&nbsp;The&nbsp;result&nbsp;is&nbsp;the&nbsp;same:&nbsp;program&nbsp;crashed&nbsp;with&nbsp;'FATAL:&nbsp;exception&nbsp;not&nbsp;rethrown'&nbsp;even&nbsp;though&nbsp;I&nbsp;faithfully&nbsp;rethrow&nbsp;exception.&nbsp;It&nbsp;is&nbsp;worth&nbsp;noting&nbsp;that&nbsp;cancel&nbsp;point&nbsp;did&nbsp;not&nbsp;seem&nbsp;to&nbsp;work&nbsp;and&nbsp;thread&nbsp;was&nbsp;wacked&nbsp;before&nbsp;call&nbsp;to&nbsp;pthread_testcancel().&lt;/FONT&gt;&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;The&nbsp;main&nbsp;problem&nbsp;as&nbsp;I&nbsp;see&nbsp;it&nbsp;is&nbsp;this:&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;Linux&nbsp;pthreads&nbsp;crashes&nbsp;the&nbsp;process&nbsp;when&nbsp;unwinding&nbsp;stack&nbsp;of&nbsp;cancelled&nbsp;thread:&nbsp;it&nbsp;throws&nbsp;system&nbsp;(non&nbsp;C++&nbsp;exception)&nbsp;which&nbsp;crashes&nbsp;the&nbsp;program&nbsp;in&nbsp;both&nbsp;cases:&lt;/FONT&gt;&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;-&nbsp;when&nbsp;caught&nbsp;and&nbsp;rethrown&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;-&nbsp;when&nbsp;not&nbsp;caught&lt;/FONT&gt;<br>
&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;C++-pthreads&nbsp;spend&nbsp;a&nbsp;lot&nbsp;of&nbsp;time&nbsp;on&nbsp;discussing&nbsp;whether&nbsp;to&nbsp;issue&nbsp;or&nbsp;not&nbsp;&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;to&nbsp;issue&nbsp;an&nbsp;exception&nbsp;when&nbsp;thread&nbsp;is&nbsp;cancelled&nbsp;but&nbsp;does&nbsp;not&nbsp;mention&nbsp;&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;the&nbsp;fact&nbsp;that&nbsp;program&nbsp;crashes&nbsp;in&nbsp;the&nbsp;way&nbsp;outlined&nbsp;above.&nbsp;It&nbsp;is&nbsp;not&nbsp;an&nbsp;&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;issue&nbsp;of&nbsp;what&nbsp;is&nbsp;better:&nbsp;to&nbsp;throw&nbsp;or&nbsp;not&nbsp;to&nbsp;throw.&nbsp;It&nbsp;is&nbsp;simply&nbsp;that&nbsp;&lt;/FONT&gt;<br>
<br>
&lt;BR&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;pthreads&nbsp;in&nbsp;current&nbsp;state&nbsp;is&nbsp;unusable&nbsp;for&nbsp;multi-threading.&lt;/FONT&gt;<br>
&lt;/P&gt;<br>
<br>
&lt;P&gt;&lt;FONT&nbsp;SIZE=2&nbsp;FACE=&quot;Courier&nbsp;New&quot;&gt;George&nbsp;Shimanovich&lt;/FONT&gt;<br>
&lt;/P&gt;<br>
<br>
&lt;/BODY&gt;<br>
&lt;/HTML&gt;
</tt>
